### PR TITLE
476 create geography subfolder in memilio and create graph for germany

### LINKED ISSUE #476: Create geography subfolder in memilio and create graph for Germany
Files such as https://github.com/DLR-SC/memilio/blob/main/cpp/memilio/epidemiology/regions.cpp or holiday_data should be moved in a new subfolder `geography`. Furthermore, we should think of creating routines for particular graphs of German counties as it is in done in simulations in this new subfolder.

This also is related to python simulations for graphs.

### PR DESCRIPTION
**Pull request by jubicker**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
Always to be checked:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [ ] Make sure that the [pre-commit linting/style checks pass](https://github.com/DLR-SC/memilio/wiki).

If functions were changed or functionality was added:
* [ ] Tests for new functionality has been added
* [x] A local test was succesful

If new functionality was added:
* [ ] There is appropriate **documentation** of your work. (use doxygen style comments)

If new third party software is used:
* [ ] Did you pay attention to its license? Please remember to add it to the wiki after successful merging.

If new mathematical methods or epidemiological terms are used:
* [ ] Are new methods referenced? Did you provide further documentation? Has the glossary been updated? 

The following questions are addressed in the documentation if need be: 
* [ ] Developers (what did you do?, how can it be maintained?)
* [ ] For users (how to use your work?)
* [ ] For admins (how to install and configure your work?)

* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [x] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [x] There are appropriate **unit tests** and they pass.
* [x] The git history is clean and linearized for the merge request. All reviewers should squash commits and write a simple and meaningful commit message.
* [x] Coverage report for new code is acceptable. 

closes #476

### CODE REVIEW

File: cpp/simulations/2021_vaccination_sarscov2_delta_germany.cpp [hunk 1]
```diff
@@ -717,14 +593,14 @@ mio::IOResult<void> run(RunMode mode, const fs::path& data_dir, const fs::path&
         temp_date = mio::Date(2021, 6, 6);
     }
     const auto start_date   = temp_date;
-    const auto num_days_sim = 90.0;
+    const auto num_days_sim = 20.0;
```
Comment (HenrZu) [inline cpp/simulations/2021_vaccination_sarscov2_delta_germany.cpp:596]:
```suggestion
    const auto num_days_sim = 90.0;
```

I think you changed the values for testing :)

File: cpp/simulations/2021_vaccination_sarscov2_delta_germany.cpp [hunk 2]
```diff
@@ -717,14 +593,14 @@ mio::IOResult<void> run(RunMode mode, const fs::path& data_dir, const fs::path&
...
     const auto end_date     = mio::offset_date_by_days(start_date, int(std::ceil(num_days_sim)));
-    const auto num_runs     = 500;
+    const auto num_runs     = 1;
```
Comment (HenrZu) [inline cpp/simulations/2021_vaccination_sarscov2_delta_germany.cpp:598]:
```suggestion
    const auto num_runs     = 500;
```

**Review by HenrZu — CHANGES_REQUESTED** [review]
Looks good. Only some short comments.
Also, you should link the Issue with the PR :)

File: cpp/simulations/2020_npis_sarscov2_wildtype_germany.cpp [hunk 3]
```diff
@@ -595,11 +460,24 @@ create_graph(mio::Date start_date, mio::Date end_date, const fs::path& data_dir)
...
     mio::Graph<mio::osecir::Model, mio::MigrationParameters> params_graph;
-    BOOST_OUTCOME_TRY(set_nodes(params, start_date, end_date, data_dir, params_graph));
-    BOOST_OUTCOME_TRY(set_edges(data_dir, params_graph));
+    const auto& read_function = mio::osecir::read_population_data_county<mio::osecir::Model>;
+    const auto& create_graph_function =
+        mio::create_graph<mio::osecir::TestAndTraceCapacity, mio::osecir::ContactPatterns, ContactLocation,
+                          mio::osecir::InfectionState, mio::osecir::Model, mio::osecir::Parameters,
+                          decltype(read_function)>;
+
+    BOOST_OUTCOME_TRY(create_graph_function(params_graph, params, start_date, end_date, data_dir, read_function,
```
[the remark this answers is not in the document]
Comment (jubicker) [reply cpp/simulations/2020_npis_sarscov2_wildtype_germany.cpp:478]: Yes they are set by default, but however there arises an error when trying to build without setting the last two input arguments

**Comment by jubicker** [comment]
closes #476

**Review by HenrZu — APPROVED** [review]

**Comment by jubicker** [comment]
@HenrZu @mknaranja I moved the functions from epi_data.h to graph.h. Can you have another look at it?

File: cpp/memilio/geography/regions.h [hunk 4]
```diff
@@ -82,16 +84,16 @@ get_holidays(StateId state);
...
+         * @param[in] start_date start of the queried period.
+         * @param[in] end_date end of the queried period.
          * @return range of pairs of start and end dates of holiday periods, sorted by start date.
          */
 Range<std::pair<std::vector<std::pair<Date, Date>>::const_iterator, std::vector<std::pair<Date, Date>>::const_iterator>>
 get_holidays(StateId state, Date start_date, Date end_date);
 
-} // namespace de
 } // namespace regions
+
```
Comment (HenrZu) [inline cpp/memilio/geography/regions.h:96]:
```suggestion
```

delete this space

File: cpp/memilio/io/epi_data.h [hunk 5]
```diff
@@ -384,8 +384,8 @@ class VaccinationDataEntry
         auto num_vaccinations_completed = obj.expect_element("Vacc_completed", Tag<double>{});
         auto date                       = obj.expect_element("Date", Tag<StringDate>{});
         auto age_group_str              = obj.expect_element("Age_RKI", Tag<std::string>{});
-        auto state_id                   = obj.expect_optional("ID_County", Tag<regions::de::StateId>{});
-        auto county_id                  = obj.expect_optional("ID_County", Tag<regions::de::CountyId>{});
+        auto state_id                   = obj.expect_optional("ID_County", Tag<regions::StateId>{});
```
Comment (HenrZu) [inline cpp/memilio/io/epi_data.h:387]:
```suggestion
        auto state_id                   = obj.expect_optional("ID_State", Tag<regions::StateId>{});
```

File: cpp/memilio/mobility/graph.h [hunk 6]
```diff
@@ -229,6 +240,124 @@ class Graph
     std::vector<Edge<EdgePropertyT>> m_edges;
 }; // namespace mio
 
+template <class TestNTrace, class ContactPattern, class Model, class MigrationParams, class Parameters,
```
Comment (HenrZu) [inline cpp/memilio/mobility/graph.h:243]:
```suggestion
template <class TestAndTrace, class ContactPattern, class Model, class MigrationParams, class Parameters,
```

File: cpp/memilio/mobility/graph.h [hunk 7]
```diff
@@ -229,6 +240,124 @@ class Graph
...
+
+    BOOST_OUTCOME_TRY(read_func(counties, start_date, county_ids, scaling_factor_inf, scaling_factor_icu,
+                                (data_dir / "pydata" / "Germany").string(), num_days, export_time_series));
+
+    for (size_t county_idx = 0; county_idx < counties.size(); ++county_idx) {
+
+        auto tnt_capacity = counties[county_idx].populations.get_total() * tnt_capacity_factor;
+
+        //local parameters
+        auto& tnt_value = counties[county_idx].parameters.template get<TestNTrace>();
```
Comment (HenrZu) [inline cpp/memilio/mobility/graph.h:265]:
```suggestion
        auto& tnt_value = counties[county_idx].parameters.template get<TestAndTrace>();
```

File: cpp/models/ode_secir/parameters_io.h [hunk 8]
```diff
@@ -318,11 +319,13 @@ IOResult<void> read_population_data_state(std::vector<Model>& model, Date date,
  * @param scaling_factor_inf factors by which to scale the confirmed cases of rki data
  * @param scaling_factor_icu factor by which to scale the icu cases of divi data
  * @param dir directory of files
+ * @param num_days Number of days to be simulated; required to extrapolate real data
```
Comment (HenrZu) [inline cpp/models/ode_secir/parameters_io.h:322]:
```suggestion
 * @param num_days [Default: 0] Number of days to be simulated; required to extrapolate real data
```

File: cpp/models/ode_secir/parameters_io.h [hunk 9]
```diff
@@ -318,11 +319,13 @@ IOResult<void> read_population_data_state(std::vector<Model>& model, Date date,
...
+ * @param export_time_series If true, reads data for each day of simulation and writes it in the same directory as the input files.
```
Comment (HenrZu) [inline cpp/models/ode_secir/parameters_io.h:323]:
```suggestion
 * @param export_time_series [Default: false] If true, reads data for each day of simulation and writes it in the same directory as the input files.
```

**Review by HenrZu — CHANGES_REQUESTED** [review]
Looks good 👍

**Review by HenrZu — APPROVED** [review]
Thanks. As discussed, the code coverage will be improved in issue #525.

**Pull request merged**